src/openms/source/QC/Suitability.cpp

```diff
@@ -0,0 +1,316 @@
+      double second_xscore_mw = double(second_hit->getMetaValue("MS:1002252")) / second_hit->getSequence().getMonoWeight();
+      if (top_xscore_mw - second_xscore_mw <= d.cut_off)
+      {
+        ++d.num_top_db;
+        ++d.num_re_ranked;
+      }
+      else
+      {
+        ++d.num_top_novo;
+      }
+    }
+
+    d.suitability = double(d.num_top_db) / (d.num_top_db + d.num_top_novo);
+
+    results_.push_back(d);
+  }
+
+  std::vector<Suitability::SuitabilityData> Suitability::getResults() const
+  {
+    return results_;
+  }
+
+  double Suitability::getDecoyDiff_(const PeptideIdentification& pep_id)
+  {
+    double diff = DBL_MAX;
+
+    // get the score of the first two decoy hits
+    double decoy_1 = DBL_MAX;
+    double decoy_2 = DBL_MAX;
+    UInt curr_hit = 0;
+
+    for (const auto& hit : pep_id.getHits())
+    {
+      if (curr_hit > 10) break;
+      ++curr_hit;
+
+      if (!hit.metaValueExists("target_decoy"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No target/decoy information found! Make sure 'PeptideIndexer' is run beforehand."));
+      }
+
+      if (!hit.metaValueExists("MS:1002252"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No cross correlation score found at peptide hit. Only Comet search engine is supported right now."));
+      }
+
+      if (decoy_1 == DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_1 = hit.getMetaValue("MS:1002252");
+        continue;
+      }
+      if (decoy_1 < DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_2 = hit.getMetaValue("MS:1002252");
+        break;
+      }
+    }
+
+    if (decoy_2 < DBL_MAX) // if there are two decoy hits
+    {
+      diff = abs(decoy_1 - decoy_2) / pep_id.getHits()[0].getSequence().getMonoWeight(); // normalized by mw
+    }
+
+    // if there aren't two decoy hits DBL_MAX is returned
+    return diff;
+  }
+
+  double Suitability::getDecoyCutOff_(const vector<PeptideIdentification>& pep_ids, double novo_fract)
+  {
+    // get all decoy diffs of peptide ids with at least two decoy hits
+    vector<double> diffs;
+    for (const auto& pep_id : pep_ids)
+    {
+      double diff = getDecoyDiff_(pep_id);
+      if (diff < DBL_MAX)
+      {
+        diffs.push_back(diff);
+      }
+    }
+
+    if (double(diffs.size()) / pep_ids.size() < 0.2)
+    {
+      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Under 20 % of peptide identifications have two decoy hits. This is not enough for re-ranking. Use the 'force_no_re_rank' flag to still compute a suitability score."));
+    }
+
+    // sort the diffs decreasing and get the (1-novo_fract)*N one
+    auto sort_end = diffs.begin() + (1 - novo_fract) * diffs.size();
+
+    partial_sort(diffs.begin(), sort_end + 1, diffs.end(), greater<double>());
```

Comment:
even better=faster should be https://en.cppreference.com/w/cpp/algorithm/nth_element.